efficiency improvements for InvSysConnector

AartBluestoke opened the pull request:
Don't build the item list for unwatched connections. 
Don't do an N^2 search over the inventory and wanted list each item.

AartBluestoke committed:
Please don't send the packet for unwatched changes

updateContentListener is called for every recieved item.

AartBluestoke committed:
overloads caused by lots of 1 item shouldn't lag

destination is now a hashmap by itemIdentifier, storing a list of all
items flowing by that id.

this means we don't need to search the list for each item.

also, if this list is empty, don't bother checking to see if each item
is in it.

slight increase in the cost of single items flowing, creating then
destroying individual almost empty lists.

massive decrease in cost of an overloaded connection.

davboecki commented:
Wouldn't it be better to use the ItemRoutingInformation(without ItemIdentifierStack) to save the destination information instead of that Triplet.
Trough using that Triplet we not only lose the JamList but also AdditionalTarget Information which are needed to handle lost items while crafting.

AartBluestoke commented:
I agree, I'll add another commit to that pr this afternoon with that change.
it should still be indexed by own identifier as most if the code says 'get
the deliveries for this item type' , and in the case of an overload it is
probably with lots of a specific item, rather than handling a few different
items in the most evident easy possible.

Also, how could that list get huge? It should be possible to be bigger than
the size of the attached inventory (an ender chest in this case) .
Also, I think we may need an 'and destination is powered' check somewhere,
as non powered pipes still 'express interest' , and items going in one side
there would indicate that something on the other side claimed it could
recurve the item even though it wasn't powered.
On 9 Oct 2014 05:46, "davboecki" notifications@github.com wrote:

> Wouldn't it be better to use the ItemRoutingInformation(without
> ItemIdentifierStack) to save the destination information instead of that
> Triplet.
> Trough using that Triplet we not only lose the JamList but also
> AdditionalTarget Information which are needed to handle lost items while
> crafting.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RS485/LogisticsPipes/pull/468#issuecomment-58415796.

AartBluestoke committed:
don't use a triplet, store the actual ItemRoutingInformation

AartBluestoke committed:
stack "might" be null if you've sent it

AartBluestoke committed:
handle partial insert into a full InvSysConnecter

LPTravelingItemServer holds a stack.
reverseItem takes only a LPTIS, and not an extra stack, which is a clone
from it.
Added comments to indicate which functions take ownership of the stack
passed into it.

itemStack clone of ArrivingItem removed from code, as nothing takes
ownership of it, and passing around things which may or may not be
duplicate or different is just asking for trouble.

when we call InsertedItemStack on a partial insert, duplicate the info
and notify of insert of the partial stack, bounce the rest.

inserted does not take a redundant item and info, only an info (updated
to hold the correct stack size)

handleItemEnterInv removes old addItem code

PipeItemsInvSysConnector::
addItem now just takes an info, specifying both the item and the
destination.
addItem now ignores 0 size stacks.

handleItemEnterInv now only takes an info and a tile.

IDirectRoutingConnection updated to accoutnt for changed addItem
changes.

AartBluestoke committed:
now with code fixes so it actually compiles

should be complete, but is untested (worked 2 commits back)

AartBluestoke committed:
typo fixes

now tested to work, clear backlog, not overload, and otherwise handle
full intermediate storage, and power cuts on busy networks.

AartBluestoke commented:
now ready for pull.
